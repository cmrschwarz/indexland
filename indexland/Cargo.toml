[package]
name = "indexland"

version.workspace = true
license.workspace = true
edition.workspace = true
rust-version.workspace = true
authors.workspace = true
repository.workspace = true
categories.workspace = true
keywords.workspace = true

description = "Rust Collections with Newtype Indices"

[lib]
name = "indexland"

[dependencies.indexland_derive]
workspace = true
optional = true
default-features = false

[dependencies.arrayvec]
version = "0.7.6"
optional = true
default-features = false

[dependencies.smallvec]
version = "1.14.0"
optional = true
default-features = false
features = ["const_new"]

[dependencies.indexmap]
version = "2.7.1"
optional = true
default-features = false

[dependencies.serde]
version = "1.0.218"
optional = true
default-features = false


[dev-dependencies]
derive_more = { version = "2.0.1", features = ["full"] }


[features]
default = ["std", "derive", "arrayvec"]

full = ["std", "derive", "nonmax", "arrayvec", "smallvec", "indexmap", "serde"]

derive = ["dep:indexland_derive"]

nonmax = []

disable_debuggable_nonmax = []

alloc = []

std = ["alloc", "arrayvec/std", "indexmap/std"]

smallvec = ["std", "dep:smallvec"] # smallvec v1 does not support no_std

arrayvec = ["dep:arrayvec"]

indexmap = ["alloc", "dep:indexmap"]

serde = ["dep:serde", "arrayvec/serde", "smallvec/serde", "indexmap/serde"]

[package.metadata.docs.rs]
features = ["full"]
