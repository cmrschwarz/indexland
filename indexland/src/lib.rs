#![doc = include_str!("../README.md")]
#![warn(clippy::pedantic)]
#![allow(clippy::missing_errors_doc)]
#![allow(clippy::missing_panics_doc)]
#![allow(clippy::must_use_candidate)]
#![allow(clippy::return_self_not_must_use)]
#![cfg_attr(docsrs, feature(doc_cfg))]
// nostd
#![no_std]
#[cfg(feature = "std")]
extern crate std;

#[cfg(feature = "alloc")]
extern crate alloc;

// allow this crate to refer to itself as ::indexland for macros to work
// within tests, see https://github.com/rust-lang/rust/issues/54647
extern crate self as indexland;

pub mod idx;
pub mod index_array;
pub mod index_enumerate;
pub mod index_range;

pub mod index_slice;
pub mod index_slice_index;

#[cfg(feature = "alloc")]
#[cfg_attr(docsrs, doc(cfg(feature = "alloc")))]
pub mod index_vec;

#[cfg(feature = "alloc")]
#[cfg_attr(docsrs, doc(cfg(feature = "alloc")))]
pub mod index_vec_deque;

#[cfg(feature = "nonmax")]
#[cfg_attr(docsrs, doc(cfg(feature = "nonmax")))]
pub mod nonmax;

#[cfg(feature = "arrayvec")]
#[cfg_attr(docsrs, doc(cfg(feature = "arrayvec")))]
pub mod index_array_vec;

#[cfg(feature = "smallvec")]
#[cfg_attr(docsrs, doc(cfg(feature = "smallvec")))]
pub mod index_small_vec;

#[cfg(feature = "indexmap")]
#[cfg_attr(docsrs, doc(cfg(feature = "indexmap")))]
pub mod index_hash_map;

#[cfg(feature = "indexmap")]
#[cfg_attr(docsrs, doc(cfg(feature = "indexmap")))]
pub mod index_hash_set;

// convenience exports

#[doc(inline)]
pub use crate::{
    idx::{Idx, IdxEnum, IdxNewtype},
    index_range::IndexRange,
    index_range::IndexRangeBounds,
};

#[doc(inline)]
pub use index_array::{EnumIndexArray, IndexArray};

#[doc(inline)]
pub use index_slice::IndexSlice;

#[cfg(feature = "alloc")]
#[doc(inline)]
pub use index_vec::IndexVec;

#[cfg(feature = "alloc")]
#[doc(inline)]
pub use index_vec_deque::IndexVecDeque;

#[cfg(feature = "derive")]
extern crate indexland_derive;

#[cfg(feature = "derive")]
#[cfg_attr(docsrs, doc(cfg(feature = "derive")))]
#[doc(inline)]
pub use indexland_derive::{Idx, IdxEnum, IdxNewtype};

#[cfg(feature = "nonmax")]
#[doc(inline)]
pub use nonmax::NonMax;

#[cfg(feature = "arrayvec")]
#[doc(inline)]
pub use index_array_vec::IndexArrayVec;

#[cfg(feature = "smallvec")]
#[doc(inline)]
pub use index_small_vec::IndexSmallVec;

#[cfg(feature = "indexmap")]
#[doc(inline)]
pub use {index_hash_map::IndexHashMap, index_hash_set::IndexHashSet};

// re-export the utility crates that we bundle
#[cfg(feature = "arrayvec")]
#[cfg_attr(docsrs, doc(cfg(feature = "arrayvec")))]
pub use arrayvec;

#[cfg(feature = "smallvec")]
#[cfg_attr(docsrs, doc(cfg(feature = "smallvec")))]
pub use smallvec;

#[cfg(feature = "indexmap")]
#[cfg_attr(docsrs, doc(cfg(feature = "indexmap")))]
pub use indexmap;

// used in macros, not public api
#[doc(hidden)]
pub mod __private {
    use core::mem::MaybeUninit;

    /// Essentially [`std::mem::MaybeUninit::transpose`] in stable Rust. Will
    /// be removed once [maybe_uninit_uninit_array_transpose](https://github.com/rust-lang/rust/issues/96097)
    /// is stabilized.
    #[allow(clippy::needless_pass_by_value)]
    pub const unsafe fn transpose_assume_uninit<T, const N: usize>(
        v: [MaybeUninit<T>; N],
    ) -> [T; N] {
        let mut res = MaybeUninit::<[T; N]>::uninit();
        let mut i = 0;
        while i < v.len() {
            unsafe {
                res.as_mut_ptr()
                    .cast::<T>()
                    .add(i)
                    .write(v.as_ptr().add(i).read().assume_init());
            };
            i += 1;
        }
        unsafe { res.assume_init() }
    }
    // reexport for vec!
    pub extern crate alloc;
}
