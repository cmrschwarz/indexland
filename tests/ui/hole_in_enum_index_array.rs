use indexland::{index_array, IndexArray};

fn main() {
    let arr: IndexArray<Foo, _, 3> = index_array! {
        0 => 2,
        2 => 3,
    };
}
