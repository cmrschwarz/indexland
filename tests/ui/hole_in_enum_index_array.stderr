error[E0412]: cannot find type `Foo` in this scope
 --> tests/ui/hole_in_enum_index_array.rs:4:25
  |
4 |     let arr: IndexArray<Foo, _, 3> = index_array! {
  |                         ^^^ not found in this scope

error[E0308]: mismatched types
 --> tests/ui/hole_in_enum_index_array.rs:4:38
  |
4 |       let arr: IndexArray<Foo, _, 3> = index_array! {
  |  ______________________________________^
5 | |         0 => 2,
6 | |         2 => 3,
7 | |     };
  | |     ^
  | |     |
  | |_____expected an array with a size of 3, found one with a size of 2
  |       arguments to this function are incorrect
  |
  = note: expected array `[_; 3]`
             found array `[{integer}; 2]`
note: associated function defined here
 --> src/index_array.rs
  |
  |     pub const fn new(data: [T; N]) -> Self {
  |                  ^^^
  = note: this error originates in the macro `index_array` (in Nightly builds, run with -Z macro-backtrace for more info)
